add tutorial configuring babel6 for node.js

abdulhannanali opened the pull request:
shrot tutorial for getting up and running with babel6 for node.js added

abdulhannanali commented:
Hi @UltCombo I have created a pull request. Thank you for this.

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+
+If yes, you can use the newer features of JavaScript ES6 and ES7 standards in your node applications today. ES6 and ES7 make the JavaScript development a cool breeze. But hey, not every ES6 feature is supported in our beloved [Node.js](https://nodejs.org).
+
+This is where [BabelJS](https://babeljs.io) comes to the rescue. BabelJS is a transpiler for JavaScript which transpiles your ES6 and ES7 code into ES5 and even ES3 code. In simple words it converts it into JavaScript that node.js can run and make you really happy.
```
In some parts of the tutorial, you call it "BabelJS" and, in others, "babel". It would be nice to have consistent naming. I think it would be better to always name it "Babel", as that is how it is written in the https://babeljs.io/ site.

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+This is where [BabelJS](https://babeljs.io) comes to the rescue. BabelJS is a transpiler for JavaScript which transpiles your ES6 and ES7 code into ES5 and even ES3 code. In simple words it converts it into JavaScript that node.js can run and make you really happy.
+<!--more-->
+
+**Little Notice:** If you just want ES6 features and don't want to babelify stuff. You can use `--harmony flag` before running your node application. In order to access more harmony flags for staging and experimental features run this command `node --v8-options | grep harmony` . But there aren't even all the features present in Node. So you may continue if you can access ES6 but also need to acce ...
```
Small nit: maybe change "`--harmony flag`" to "`--harmony` flag"

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+### Following along the code
+Type of person who follows code instead of just reading? Code is available here in this [repo](https://github.com/abdulhannanali/babel-configuration-tutorial)
+
+### Installing and getting started with babeljs
```
Same as https://github.com/JSRocksHQ/jsrockshq.github.io/pull/72/files#r48644774
babeljs -> Babel

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+Type of person who follows code instead of just reading? Code is available here in this [repo](https://github.com/abdulhannanali/babel-configuration-tutorial)
+
+### Installing and getting started with babeljs
+There's not a single way you will be able to setup Babel. Here we will be discussing enough to get up and running with it.
```
"There's not a single way you will be able to setup Babel."
I think this statement could be misinterpreted as "There is no way you will be able to setup Babel.". Perhaps change "There's not a single way" to "There is more than one way".

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
````diff
@@ -0,0 +1,164 @@
+```bash
+npm install --save-dev babel-preset-es2015 babel-preset-stage-0
+```
+In order to access a wide range of babel plug ins click [here](https://babeljs.io/docs/plugins/)
````
babel -> Babel
plug ins -> plugins

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
````diff
@@ -0,0 +1,164 @@
+```
+In order to access a wide range of babel plug ins click [here](https://babeljs.io/docs/plugins/)
+
+You need to include these presets in command you issue
````
in command -> in the command (feels better to read, IMO)

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
````diff
@@ -0,0 +1,164 @@
+node build/app.js
+```
+
+### Setting up a proper build environment using babel
````
babel -> Babel

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+### Setting up a proper build environment using babel
+This is all good magic but what about doing some proper development using node.js.
+
+#### babel configuration file .babelrc
```
babel -> Babel

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
````diff
@@ -0,0 +1,164 @@
+babel --presets es2015,stage-0 code/index.js -o build/app.js
+```
+
+You will see normal es5 code generated in `app.js`, this is called **Transpiled code** (a term used widely in JS World). You can run this code using the command below.
````
suggestion: a term used widely in JS World -> a term widely used in the JS world

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+This is all good magic but what about doing some proper development using node.js.
+
+#### babel configuration file .babelrc
+`.babelrc` is a very neat way to separate all your babel stuff in one JSON file. It's also pretty easy to get started. This is our .babelrc file for this tutorial. You can access other .babelrc options [here](http://babeljs.io/docs/usage/options/) and make it as complex as you want it to be
```
babel -> Babel
I guess "complex" is not a positive word in this context, perhaps "complete", "robust" or something else would be more appropriate.

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
````diff
@@ -0,0 +1,164 @@
+  "plugins": ["es2015", "stage-0"]
+}
+```
+This is pretty much it for this tutorial, this will help us with our rest of development. Now whenever we want to add or subtract plugins. Instead of changing the command we will change the plugins array in this file. Easy! Isn't it?
````
"this will help us with our rest of development" sounds a bit odd to me. I'm not really sure what you meant—did you mean that the Babel setup improves developer experience? I'm not sure what "this" and "rest of development" refers to in that statement.

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
````diff
@@ -0,0 +1,164 @@
+```js
+errorGen.next()
+         ^
+Error: source maps are awesome
````
I think it would be nice if you added the error's call stack as well—that's basically the main point of source maps, making the call stack point to the lines in the original sources instead of the transpiled ones.

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+So this is how you'll use source maps
+
+#### Setting up npm command
+In order to simplify the build process even more. You can update your `package.json` file to include a build script for babel. In `package.json` script object you can add a build script such as the one below
```
I think a comma is more suitable than a full stop here: "In order to simplify the build process even more, you can update your `package.json` file to include a build script for Babel."
babel -> Babel

abdulhannanali commented:
@UltCombo thank you for your time to correct this tutorial. Yup! It meant that this will improve the developer experience but I am going to remove this line.  Thank you for pointing it out

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+
+
+#### Learn more about Babel
+This is a basic tutorial on babel but the Babel world just starts here. It's surrounded by a wonderful community and is used by big names in IT world. Babel has support for all the major build tools too such as [GruntJS](https://www.npmjs.com/package/grunt-babel) and [GulpJS](https://npmjs.org/package/gulp-babel/). You can check them all out here on [Babel Website](https://babeljs.io/docs/setup/)
```
babel -> Babel.
Some nits:
- GruntJS -> Grunt (as written in [their website](http://gruntjs.com/))
- GulpJS -> gulp ([gulp is always written lowercase](https://github.com/gulpjs/gulp/blob/master/docs/FAQ.md#is-it-gulp-or-gulp))

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+#### Learn more about Babel
+This is a basic tutorial on babel but the Babel world just starts here. It's surrounded by a wonderful community and is used by big names in IT world. Babel has support for all the major build tools too such as [GruntJS](https://www.npmjs.com/package/grunt-babel) and [GulpJS](https://npmjs.org/package/gulp-babel/). You can check them all out here on [Babel Website](https://babeljs.io/docs/setup/)
+
+These are some of the resources that can even up your game further in the babel world
```
babel -> Babel

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,164 @@
+
+If you find some typo or would like to make some update. Please do so using the power of issues and PR in our [Github Repo](https://github.com/abdulhannanali/babel-configuration-tutorial).
+
+I would also like to Thank [Fabrício Matté](http://ultcombo.js.org/) for approving this article to be posted on JSRocks and the corrections he made.
```
Thank -> thank
JSRocks -> JS Rocks (that is how it appears in the site title)
Also, I don't think the statement is entirely necessary, but you can leave it if you like. :smile:

UltCombo commented:
Excellent work!
I added some review comments above, as you've already noticed.
And sorry, I got a bit carried away with the comments hahah. I think it would be faster if I fixed the typos myself when pulling this in, but I didn't want to change too much without your acknowledging it.

A couple side-notes:
- We usually prefer writing "ES2015"/"ES2016" over "ES6"/"ES7," as the former are more up-to-date names. Both namings are pretty well know, so we can leave this to the author's preference.
- Our code style uses semicolons to terminate statements, but again, your code is valid so we can leave it to the author's preference.

I didn't want to be too picky as that drives contributors away, so you don't have to change what I noted above (they are just side-notes). We can leave it to the author's (your) preference. :smile:

abdulhannanali commented:
Thanks @UltCombo That is so awesome of you! Thank you for correcting so many mistakes. I have also renamed ES6/ES7 to ES2015/ES2016 and added semi colons. Also thank you for pointing out the mistakes and let me correct them. They surely improved my knowledge.

abdulhannanali commented:
You are doing right by maintaining good standards in JS Rocks. This will surely lead to high quality content on the website.

abdulhannanali commented:
@UltCombo As per your suggestions, I have made some changes. Please tell me if there are more changes to make anywhere else too.

UltCombo commented:
> Thanks @UltCombo That is so awesome of you! Thank you for correcting so many mistakes. I have also renamed ES6/ES7 to ES2015/ES2016 and added semi colons. Also thank you for pointing out the mistakes and let me correct them. They surely improved my knowledge.

No problem, you're welcome. :smiley:

> @UltCombo As per your suggestions, I have made some changes. Please tell me if there are more changes to make anywhere else too.

Thank you for your hard work and promptly response to my reviews. :smile:
This seems ready to go, I'll just read through it one more time and fix any typos I find to save us some time, then it should be ready for publication. ;)

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,175 @@
+This is where [Babel](https://babeljs.io) comes to the rescue. Babel is a transpiler for JavaScript which transpiles your ES2015 and ES2016 code into ES5 and even ES3 code. In simple words it converts it into JavaScript that node.js can run and make you really happy.
+<!--more-->
+
+**Little Notice:** If you just want ES2015 features and don't want to Babelify stuff. You can use `--harmony` flag before running your node application. In order to access more harmony flags for staging and experimental features run this command `node --v8-options | grep harmony` but there aren't even all the features present in Node. So you may continue if you can access ES2015 but also need to  ...
```
Seems like there are a few issues in this paragraph.

- The current version of Node supports most ES2015 features without requiring the harmony flag.
- The flagged features are often incomplete or otherwise broken. I think it would be good to note this.

@abdulhannanali What do you think about this revised suggestion:

**Small note:** Recent versions of Node.js already support several ES2015 features, so if you don't like transpiling you can use the [`--harmony` flag](https://nodejs.org/en/docs/es6/) when running your Node.js application to a few more (staged) features. In order to access more harmony flags for staging and experimental features run this command: `node --v8-options | grep harmony`. However, note that not all features are supported even in the latest Node.js (version 5 at the time of writing), and the flagged features are often unstable or incomplete. So you may keep reading if you want to make use of more ES2015 and **ES2016** features without adding those pesky flags.

Maybe this is too complex/too much info in a paragraph, not sure.

UltCombo replied:
Revised my revision:

**Small note:** Recent versions of Node.js already support several ES2015 features, so if you don't like transpiling you can use the [`--harmony` flag](https://nodejs.org/en/docs/es6/) when running your Node.js application to enable a few more (staged) features. In order to access more harmony flags for experimental features run this command: `node --v8-options | grep harmony`. However, note that not all features are supported even in the latest Node.js (version 5 at the time of writing), and the flagged features are often unstable or incomplete. So keep reading if you want to make use of more ES2015 and **ES2016** features without adding those pesky flags.

@abdulhannanali What do you think?

UltCombo replied:
I still think we can make this shorter, while covering the points I've noted earlier.

UltCombo replied:
One more try:

**Small note:** Node.js already supports several ES2015 features, so if you dislike transpiling you may run Node.js with the [`--harmony` flag](https://nodejs.org/en/docs/es6/) to enable a few more (staged) features. In order to access more harmony flags for experimental features run this command: `node --v8-options | grep harmony`. However, note that not all features are supported even in the latest Node.js (version 5 at the time of writing), and the flagged features are often unstable or incomplete. So keep reading to make use of more ES2015 and **ES2016** features, without requiring flags.

UltCombo replied:
I think the revision above is good. @abdulhannanali Do you think it is okay?

abdulhannanali replied:
@UltCombo Last revision is fine, what about just removing the word dislike and replace it with some thing more generic such as don't want to cos in some cases this is not the matter of liking or disliking. I wrote it but I am asking for suggestion. Also thanks for correcting and updating this :)

UltCombo replied:
"don't want to" sounds good. :+1:

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,175 @@
+It will read the **presets** to use from `.babelrc` compile the code in `code/` directory and generate the compiled code javascript files in `build/` folder but hey! The command didn't end. Notice the `-w` flag, **w** stands for **watch**, it will recompile the code as you make changes in your code directory. COOL! Now this is some magic I am talking about.
+
+#### Using Source maps in your file
+If you are thinking that's all cool and fun but what about syntax highlighting and real time code debugging. You don't have to be worried. Source maps are just for this purpose. Source maps tell node.js that this code is transpiled and let's you highlight **source** file instead of the **transpiled** file 😄. YAYY!
```
I guess source maps don't have much to do with "syntax highlighting" (when I read "syntax highlighting" I think about a text editor's syntax highlighting, not sure if this is what you mean).

> let's you highlight **source** file instead of the **transpiled** file

I guess there's a better verb than "highlight" here. Maybe "inspect" or "debug".

UltCombo replied:
let's -> lets

abdulhannanali replied:
@UltCombo Surely! They don't help us in syntax highlighting that depends on IDE. I'll make a PR about it.

UltCombo commented:
I've added a few more comments above. It is getting a bit late here, so after we resolve those nits I can do one final review round tomorrow and we can finally publish your tutorial. Thanks for the taking the time to contribute. :smile:

abdulhannanali commented:
@UltCombo Thanks for taking out the time to point out so many mistakes. I'll correct them in the next commit

abdulhannanali commented:
Also sorry for late response from me too.

UltCombo commented on src/posts/en/configuring-babel-6-for-node-js.md:
```diff
@@ -0,0 +1,175 @@
+##### Source code and Contributions and Thank yous
+Source code for this tutorial is available in this [repo](https://github.com/abdulhannanali/babel-configuration-tutorial).
+
+If you find some typo or would like to make some update. Please do so using the power of issues and PR in our [Github Repo](https://github.com/abdulhannanali/babel-configuration-tutorial).
```
full stop -> comma
You can add a link to the [JS Rocks repo](https://github.com/JSRocksHQ/jsrockshq.github.io) too if you'd like.

UltCombo commented:
@abdulhannanali 

> Also sorry for late response from me too.

No problem, we are in different time zones and life has more than just open source. :stuck_out_tongue_winking_eye:

> I'll correct them in the next commit

Thanks, after this review round we can go ahead and publish it. :smile:

UltCombo commented:
Hey, @abdulhannanali. Please take a look at the review comments above when you have time. :smiley:

abdulhannanali commented:
@UltCombo I am really sorry for the delay! I got sick and was barely able to do the codein tasks along with my sickness. I will surely take a look today.

abdulhannanali commented:
Thanks @UltCombo for reminding me. It was in my mind all this time

UltCombo commented:
@abdulhannanali No problem, get well soon. No hurries either, take it at your own pace. :smiley:

abdulhannanali commented:
@UltCombo i've made a few corrections, sorry for the long delay, tell me if it's good to go other wise we can always make more corrections. :smile: also thank you for being so much patient :+1:

UltCombo commented:
That looks awesome! I'll just fix a few typos to save you the trouble and publish it right away. :smile:

UltCombo commented:
I've made a few adjustments here: 8344aa748245d7fd22e5054c975338a76619dd52. Feel free to comment or send PRs to improve. :smiley:

@abdulhannanali Thanks again for contributing, and congratz on your first JS Rocks post! :tada: It is live here: http://jsrocks.org/2015/12/configuring-babel-6-for-node-js/

@jaydson Can you schedule an [@es6rocks](https://twitter.com/es6rocks) tweet? Maybe for tomorrow?

UltCombo commented:
Oh damn it, I forgot to update the post date. The link now rightfully points to 2016/01: http://jsrocks.org/2016/01/configuring-babel-6-for-node-js/

hacke2 commented:
wonderful! I will translate to chinese in this week

UltCombo commented:
Thanks @hacke2! :+1:

UltCombo commented:
By the way, I've committed a few more improvements so here is the [updated diff](https://github.com/JSRocksHQ/jsrockshq.github.io/compare/495b532b94fd94b3d335febc3d0618305a9b19b8...801721a90beb2759af30bf79f92cc68b38dce8eb?expand=1).

abdulhannanali commented:
@UltCombo thank you for your patience and kind help.. :)

abdulhannanali commented:
@UltCombo can we still make a few changes? there are some mistakes that are really necessary and must be corrected.

abdulhannanali commented:
i am making a pull request

UltCombo commented:
Of course. Feel free to send PRs whenever you want. :wink:
JS Rocks is a collaborative blog, so others can send suggestions and improvements as well.

UltCombo commented:
Hey, @abdulhannanali.
Are there still corrections to be made? Don't be afraid to ask for help if you need. :wink:
